### PR TITLE
Adds force press gesture detector and recognizer

### LINKED ISSUE #4838: Expose "3D Touch" handlers in GestureDetector
iOS has a "3D touch" feature. We could allow developers to respond to a 3D touch from within a Flutter app.

### PR DESCRIPTION
**Pull request by jslavitz**
Adds for force touch support for force touch enabled devices through the creation of a force touch gesture recognizer. Fixes #4838.

### CODE REVIEW

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+      final double pressure = _inverseLerp(event.pressureMin, event.pressureMax, event.pressure);
+      if (_state == _ForceState.possible) {
+        _lastTimeStamp = event.timeStamp;
+        _lastPosition = event.position;
+        if (pressure > startPressure) {
+          _state = _ForceState.started;
+          resolve(GestureDisposition.accepted);
+        } else if (event.delta.distanceSquared > kTouchSlop)
+          resolve(GestureDisposition.rejected);
+      } else if (_state == _ForceState.accepted || _state == _ForceState.peaked || _state == _ForceState.started) {
```
Comment (xster) [inline packages/flutter/lib/src/gestures/force_press.dart]: Having 2 layers of long conditionals here and in the next line is hard to read. 

Maybe just deal with if possible, then return since I don't think it's possible to do more in the same call.

Then add a few sequences of ifs so they're not nested.

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+              sourceTimeStamp: event.timeStamp,
+              pressure: pressure,
+              globalPosition: event.position,
+            )));
+          }
+        }
+      }
+    }
+    stopTrackingIfPointerNoLongerDown(event);
+    _lastTimeStamp = event.timeStamp;
```
Comment (xster) [inline packages/flutter/lib/src/gestures/force_press.dart]: not clear why this is present twice in this function. Don't you want to do this regardless?

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    _lastPosition = event.position;
+  }
+
+  void _gestureStarted() {
+    final Duration timestamp = _lastTimeStamp;
+    final Offset position = _lastPosition;
+    _lastTimeStamp = null;
```
Comment (xster) [inline packages/flutter/lib/src/gestures/force_press.dart]: Not sure why this is needed. Might need a code comment.

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+          }
+        }
+      }
+    }
+    stopTrackingIfPointerNoLongerDown(event);
+    _lastTimeStamp = event.timeStamp;
+    _lastPosition = event.position;
+  }
+
+  void _gestureStarted() {
```
Comment (xster) [inline packages/flutter/lib/src/gestures/force_press.dart]: I would keep the same pattern. Either do everything inline in handleEvent or make all callbacks their own functions.

File: packages/flutter/test/widgets/gesture_detector_test.dart [hunk 5]
```diff
@@ -343,4 +343,67 @@ void main() {
     await longPress(kLongPressTimeout + const Duration(seconds: 1)); // To make sure the time for long press has occurred
     expect(longPressUp, 1);
   });
+
```
Comment (xster) [inline packages/flutter/test/widgets/gesture_detector_test.dart]: I would add tests for interactions too, to check for instance that a long press then forced press doesn't trigger the force press callbacks.

**Review by xster — COMMENTED** [review]
Thanks!

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+
+enum _ForceState {
+  ready,
+  possible,
+  accepted,
+  started,
+  peaked,
+}
+
+/// Details object for callbacks that use [GestureForcePressStartCallback],
```
[the remark this answers is not in the document]
Comment (jslavitz) [reply packages/flutter/lib/src/gestures/force_press.dart]: Yes? But the pressure value will be sent in onUpdate. In this way, onStart and the first onUpdate will sent in the same frames. It's the way it's done for monodrag.dart so I thought it would make sense to replicate the behavior on this recognizer.

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+typedef GestureForcePressEndCallback = void Function(ForcePressDetails details);
+
+/// Recognizes a force press on devices that have force sensors.
+///
+/// Only the force from a single pointer is used to invoke events.
+class ForcePressGestureRecognizer extends OneSequenceGestureRecognizer {
+  /// Creates a force press gesture recognizer.
+  ///
+  /// The [startPressure] defaults to 0.4, and [peakPressure] defaults to 0.85
+  /// where a value of 0.0 is no pressure and a value of 1.0 is maximum pressure.
```
[the remark this answers is not in the document]
Comment (jslavitz) [reply packages/flutter/lib/src/gestures/force_press.dart]: So, I don't know if this is the wrong approach, but the events that come in off the device actually have pressures in the range of [0, 6.66], however, I interpolate the recorded values into the range [0,1], so that functionality will work the same across all devices even in this range is different. Maybe it would make more sense to change the default pointer pressure max on these events to 6.66?

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+  ForcePressGestureRecognizer({
+    this.startPressure = 0.4,
+    this.peakPressure = 0.85,
+    Object debugOwner,
+  }) : assert(startPressure != null),
+       assert(peakPressure != null),
+       super(debugOwner: debugOwner);
+
+  /// A pointer is in contact with the screen and has just pressed with a force
+  /// exceeding the [startPressure].
```
[the remark this answers is not in the document]
Comment (jslavitz) [reply packages/flutter/lib/src/gestures/force_press.dart]: Are you referring to an onDown callback? Similar to the onDown in the monodrag detector?

File: packages/flutter/lib/src/gestures/force_press.dart (hunk 1 above)
Comment (jslavitz) [reply packages/flutter/lib/src/gestures/force_press.dart]: So the only reason not to return is if we want to call onUpdate the same frame as onStart. Which seems to be a good move? Would it be ok to just separate them into two different ifs but not return?

File: packages/flutter/lib/src/gestures/events.dart [hunk 9]
```diff
@@ -334,8 +334,9 @@ class PointerAddedEvent extends PointerEvent {
     int device = 0,
     Offset position = Offset.zero,
     bool obscured = false,
-    double pressureMin = 1.0,
-    double pressureMax = 1.0,
+    double pressure = 1.0,
+    double pressureMin = 0.0,
+    double pressureMax = 6.0,
```
Comment (Hixie) [inline packages/flutter/lib/src/gestures/events.dart]: what is 6.0? This seems like a very arbitrary number.

File: packages/flutter/lib/src/gestures/events.dart [hunk 10]
```diff
@@ -334,8 +334,9 @@ class PointerAddedEvent extends PointerEvent {
     int device = 0,
     Offset position = Offset.zero,
     bool obscured = false,
-    double pressureMin = 1.0,
-    double pressureMax = 1.0,
+    double pressure = 1.0,
```
Comment (Hixie) [inline packages/flutter/lib/src/gestures/events.dart]: PointerAdded shouldn't have a pressure, there's no interaction by the point this event is sent.

File: packages/flutter/lib/src/gestures/events.dart (hunk 9 above)
Comment (Hixie) [reply packages/flutter/lib/src/gestures/events.dart]: This default range seems very wrong to me. If you create a default PointerAdded event, it should have the most conservative values. 0-6 doesn't make sense e.g. for a mouse (which should be 1.0-1.0).

File: packages/flutter/lib/src/gestures/events.dart [hunk 11]
```diff
@@ -383,11 +386,12 @@ class PointerRemovedEvent extends PointerEvent {
     device: device,
     position: null,
     obscured: obscured,
+    pressure: pressure,
     pressureMin: pressureMin,
     pressureMax: pressureMax,
     distanceMax: distanceMax,
     radiusMin: radiusMin,
-    radiusMax: radiusMax
+    radiusMax: radiusMax,
```
Comment (Hixie) [inline packages/flutter/lib/src/gestures/events.dart]: same comments here

File: packages/flutter/lib/src/gestures/events.dart [hunk 12]
```diff
@@ -410,8 +414,9 @@ class PointerHoverEvent extends PointerEvent {
     Offset delta = Offset.zero,
     int buttons = 0,
     bool obscured = false,
-    double pressureMin = 1.0,
-    double pressureMax = 1.0,
+    double pressure = 0.0,
+    double pressureMin = 0.0,
+    double pressureMax = 6.0,
```
Comment (Hixie) [inline packages/flutter/lib/src/gestures/events.dart]: this seems wrong too. By definition a hover event can't have any pressure. Surely all three numbers should be 1.0 (like they were before).

File: packages/flutter/lib/src/gestures/events.dart [hunk 13]
```diff
@@ -460,8 +466,8 @@ class PointerDownEvent extends PointerEvent {
     int buttons = 0,
     bool obscured = false,
     double pressure = 1.0,
-    double pressureMin = 1.0,
-    double pressureMax = 1.0,
+    double pressureMin = 0.0,
+    double pressureMax = 6.0,
```
Comment (Hixie) [inline packages/flutter/lib/src/gestures/events.dart]: why would the defaults change?

File: packages/flutter/lib/src/gestures/events.dart [hunk 14]
```diff
@@ -616,8 +622,9 @@ class PointerCancelEvent extends PointerEvent {
     Offset position = Offset.zero,
     int buttons = 0,
     bool obscured = false,
-    double pressureMin = 1.0,
-    double pressureMax = 1.0,
+    double pressure = 1.0,
+    double pressureMin = 0.0,
+    double pressureMax = 6.0,
```
Comment (Hixie) [inline packages/flutter/lib/src/gestures/events.dart]: cancel and up don't seem like they could have a pressure either.

File: packages/flutter/lib/src/gestures/events.dart [hunk 15]
```diff
@@ -636,6 +643,7 @@ class PointerCancelEvent extends PointerEvent {
     buttons: buttons,
     down: false,
     obscured: obscured,
+    pressure: pressure,
     pressureMin: pressureMin,
     pressureMax: pressureMax,
     distance: distance,
```
Comment (Hixie) [inline packages/flutter/lib/src/gestures/events.dart]: i don't understand any of the changes in this file.

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+///
+/// Only the force from a single pointer is used to invoke events. A tap
+/// recognizer will win against this recognizer on pointer up as long as the
+/// pointer has not pressed with a force greater than
+/// [ForcePressGestureRecognizer.startPressure]. A long press recognizer will
+/// win when the press down time exceeds the threshold time as long as the
+/// pointer's pressure was never greater than
+/// [ForcePressGestureRecognizer.startPressure] in that duration.
+///
+/// As of 11-21-18 only iPhone 6S devices and higher numbered devices have force
```
Comment (Hixie) [inline packages/flutter/lib/src/gestures/force_press.dart]: If you want to include a date please use the format "YYYY-MM-DD" or "Month YYYY", otherwise the date is wildly ambiguous.

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 17]
```diff
@@ -0,0 +1,255 @@
...
+  ///
+  /// The position of the pointer is provided in the callback's `details`
+  /// argument, which is a [ForcePressDetails] object.
+  GestureForcePressStartCallback onStart;
+
+  /// A pointer is in contact with the screen and is either moving on the plane
+  /// of the screen, pressing the screen with varying forces or both
+  /// simultaneously.
+  ///
+  /// This callback will be invoked for every frame after the invocation of
```
Comment (Hixie) [inline packages/flutter/lib/src/gestures/force_press.dart]: This mentions frames but I don't think that's accurate. Won't it be sent for every pointer event, rather than every frame? I don't see any code that tries to ensure it's frame-aligned or frame-buffered.

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 18]
```diff
@@ -0,0 +1,255 @@
...
+          _state = _ForceState.started;
+          resolve(GestureDisposition.accepted);
+        } else if (event.delta.distanceSquared > kTouchSlop)
+          resolve(GestureDisposition.rejected);
+      }
+      // In case this is the only gesture detector we still don't want to start
+      // the gesture until the pressure is greater than the startPressure.
+      if (pressure > startPressure && _state == _ForceState.accepted) {
+        _state = _ForceState.started;
+        invokeCallback<void>('onStart', () => onStart(ForcePressDetails(
```
Comment (Hixie) [inline packages/flutter/lib/src/gestures/force_press.dart]: what if onStart is null?

Comment (Hixie) [inline packages/flutter/lib/src/gestures/force_press.dart]: this implies we'll call onStart twice, once when you go from possible to accepted, and once from accepted to started.

Comment (Hixie) [reply packages/flutter/lib/src/gestures/force_press.dart]: oh never mind, i misread the code below.

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 19]
```diff
@@ -0,0 +1,255 @@
...
+      }
+    }
+    stopTrackingIfPointerNoLongerDown(event);
+  }
+
+  @override
+  void acceptGesture(int pointer) {
+    if (_state != _ForceState.accepted) {
+      _state = _ForceState.accepted;
+      if (onStart != null && _state == _ForceState.started) {
```
Comment (Hixie) [inline packages/flutter/lib/src/gestures/force_press.dart]: this is dead code, since you set _state to accepted immediately above this.

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 20]
```diff
@@ -0,0 +1,255 @@
...
+        )));
+      }
+    }
+  }
+
+  @override
+  void didStopTrackingLastPointer(int pointer) {
+    final bool wasAccepted = _state == _ForceState.started || _state == _ForceState.peaked;
+    _state = _ForceState.ready;
+    if (_state == _ForceState.possible) {
```
Comment (Hixie) [inline packages/flutter/lib/src/gestures/force_press.dart]: similarly here

File: packages/flutter/lib/src/widgets/gesture_detector.dart [hunk 21]
```diff
@@ -318,6 +326,25 @@ class GestureDetector extends StatelessWidget {
...
+  final GestureForcePressPeakCallback onForcePressPeak;
+
+  /// A pointer is in contact with the screen, has previously passed the
+  /// [ForcePressGestureRecognizer.startPressure] and is either moving on the
+  /// plane of the screen, pressing the screen with varying forces or both
+  /// simultaneously.
+  final GestureForcePressUpdateCallback onForcePressUpdate;
+
+  /// The pointer is no longer in contact with the screen.
+  final GestureForcePressEndCallback onForcePressEnd;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/gesture_detector.dart]: these should probably mention that only some devices support force presses.

File: packages/flutter/test/gestures/force_press_test.dart [hunk 22]
```diff
@@ -0,0 +1,230 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
```
Comment (Hixie) [inline packages/flutter/test/gestures/force_press_test.dart]: you need some tests to catch the issues i raised
you need some tests to test the details object

File: packages/flutter_test/lib/src/test_pointer.dart [hunk 23]
```diff
@@ -162,16 +172,24 @@ class TestGesture {
...
+                                                                             double pressureMin = 0.0,
+                                                                             double pressureMax = 1.0 }) {
     return TestAsyncUtils.guard<void>(() {
       assert(_pointer._isDown);
-      return _dispatcher(_pointer.move(location, timeStamp: timeStamp), _result);
+      return _dispatcher(_pointer.move(location, timeStamp: timeStamp, pressure: pressure,
+                                                                       pressureMin: pressureMin,
+                                                                       pressureMax: pressureMax), _result);
     });
   }
```
Comment (Hixie) [inline packages/flutter_test/lib/src/test_pointer.dart]: i think we should probably not add the pressure arguments to the TestPointer and WidgetController APIs, because that will quickly get out of hand when we add everything else also.

Better instead to add a new method to TestPointer that takes an arbitrary PointerEvent, and updates isDown accordingly, and returns it, and then a new pair of methods to TestGesture, one to create a gesture from a custom PointerDownEvent and one to send the next PointerEvent. (The existing methods can all be refactored in terms of these more generic methods too.)

That way these APIs will handle all future API changes too without having to keep growing to match the underlying pointer API.

File: packages/flutter/lib/src/gestures/events.dart (hunk 9 above)
Comment (jslavitz) [reply packages/flutter/lib/src/gestures/events.dart]: It's 6 because it's close to 6.66 which is the maximum value that apple uses for the pressure in the event itself. What value would you suggest? A mouse doesn't have any pressure by the way?

File: packages/flutter/lib/src/gestures/events.dart (hunk 15 above)
Comment (jslavitz) [reply packages/flutter/lib/src/gestures/events.dart]: Ok, let's talk about it tomorrow!

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 24]
```diff
@@ -0,0 +1,255 @@
...
+  /// The pressure of the press required to initiate a force press.
+  ///
+  /// A value of 0.0 is no pressure, and 1.0 is maximum pressure.
+  double startPressure;
+
+  /// The pressure of the press required to peak a force press.
+  ///
+  /// A value of 0.0 is no pressure, and 1.0 is maximum pressure. This value
+  /// must be greater than [startPressure].
+  double peakPressure;
```
[the remark this answers is not in the document]
Comment (jslavitz) [reply packages/flutter/lib/src/gestures/force_press.dart]: Would it maybe make more sense to make them immutable variables? I don't think it makes sense for these pressures values to be changed after the instantiation of the class.

File: packages/flutter/lib/src/gestures/force_press.dart (hunk 17 above)
Comment (jslavitz) [reply packages/flutter/lib/src/gestures/force_press.dart]: yup!

File: packages/flutter/test/gestures/force_press_test.dart (hunk 22 above)
Comment (jslavitz) [reply packages/flutter/test/gestures/force_press_test.dart]: I don't see any tests for details objects for other gesture detectors?

File: packages/flutter/lib/src/gestures/events.dart (hunk 9 above)
Comment (xster) [reply packages/flutter/lib/src/gestures/events.dart]: The default min-max should be a conservative default. If the engine sends a range from the device platform, the override should come from the engine.

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 25]
```diff
@@ -0,0 +1,262 @@
...
+
+enum _ForceState {
+  // No pointer has touched down and the detector is ready for a pointer down to occur.
+  ready,
+
+  // A pointer has touched down, but a force press gesture has not yet been detected.
+  possible,
+
+  // A pointer is down and a force press gesture has been detected. However, if
+  // ForcePressGestureRecognizer the only recognizer in the arena, the gesture
```
Comment (xster) [inline packages/flutter/lib/src/gestures/force_press.dart]: Not sure I understand this comment. What makes the gesture get accepted if it hasn't started yet?

File: packages/flutter/lib/src/gestures/force_press.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (xster) [reply packages/flutter/lib/src/gestures/force_press.dart]: ya, we're in a bit of a pickle here. Unfortunately neither solution will be very convenient. 

Linearly interpolating is pretty convenient today but the apple API makes no contractual promises with the range or the curve of the values. Neither does Android and things are even more likely to be the wild west there. Trying to shield our users from the rawness of it will likely reach a limit sooner or later. 

Just referring to some arbitrary absolute value would be crappy too since the user has to go capture the device's min/max range first before constructing this recognizer.

I think we have 2 options: 
1- Make getting the device's capability super statically easy (like https://developer.android.com/reference/android/view/InputDevice.MotionRange on Android).
2- Composability is likely our friend. Let there be 2 customizable predicate function arguments where the user can decide whether start and peak are reached given the device's min and max pressure. That function itself can actually have a const default function that maps to a reasonable value but it's overridable. 

@Hixie for opinions and prior art too.

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 26]
```diff
@@ -0,0 +1,262 @@
...
+       super(debugOwner: debugOwner);
+
+  /// A pointer is in contact with the screen and has just pressed with a force
+  /// exceeding the [startPressure]. Consequently, if there were other gesture
+  /// detectors, only the force press gesture will be detected and all others
+  /// will be rejected.
+  ///
+  /// The position of the pointer is provided in the callback's `details`
+  /// argument, which is a [ForcePressDetails] object.
+  GestureForcePressStartCallback onStart;
```
Comment (xster) [inline packages/flutter/lib/src/gestures/force_press.dart]: Suspicious that these aren't final and aren't in the constructor

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 27]
```diff
@@ -0,0 +1,262 @@
...
+
+      if (_state == _ForceState.possible) {
+        if (pressure > startPressure) {
+          _state = _ForceState.started;
+          resolve(GestureDisposition.accepted);
+        } else if (event.delta.distanceSquared > kTouchSlop)
+          resolve(GestureDisposition.rejected);
+      }
+      // In case this is the only gesture detector we still don't want to start
+      // the gesture until the pressure is greater than the startPressure.
```
Comment (xster) [inline packages/flutter/lib/src/gestures/force_press.dart]: Not sure what this means. If line 181 ran, wouldn't this always always run? Maybe I'm missing something.

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 28]
```diff
@@ -0,0 +1,262 @@
...
+  /// The position of the pointer is provided in the callback's `details`
+  /// argument, which is a [ForcePressDetails] object.
+  GestureForcePressStartCallback onStart;
+
+  /// A pointer is in contact with the screen and is either moving on the plane
+  /// of the screen, pressing the screen with varying forces or both
+  /// simultaneously.
+  ///
+  /// This callback will be invoked for every pointer event after the invocation
+  /// of [onStart] and before the invocation of [onEnd], no matter what the
```
Comment (xster) [inline packages/flutter/lib/src/gestures/force_press.dart]: and/or onPeak and before...

File: packages/flutter/lib/src/gestures/force_press.dart (hunk 26 above)
Comment (xster) [reply packages/flutter/lib/src/gestures/force_press.dart]: Seems like the other recognizers do this too. Never mind.

File: packages/flutter/test/gestures/force_press_test.dart [hunk 29]
```diff
@@ -0,0 +1,329 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
```
Comment (xster) [inline packages/flutter/test/gestures/force_press_test.dart]: Uber nit: the year is wrong

File: packages/flutter/lib/src/gestures/force_press.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (jslavitz) [reply packages/flutter/lib/src/gestures/force_press.dart]: I decided to just add the pressure into all the calls – so you don't have to have an onUpdate to get the pressure if you really only want the onStart call.

File: packages/flutter/lib/src/gestures/force_press.dart (hunk 25 above)
Comment (jslavitz) [reply packages/flutter/lib/src/gestures/force_press.dart]: So apparently gestures will be accepted by default if they're the only gestures in the arena.

File: packages/flutter/lib/src/gestures/force_press.dart (hunk 27 above)
Comment (jslavitz) [reply packages/flutter/lib/src/gestures/force_press.dart]: It wouldn't because I set the state to started before line 181 runs. I'm navigating the case where the gesture has been accepted, but the pressure isn't greater than the start pressure. That's the scenario in which this line will run.

**Comment by Hixie** [comment]
To be clear, the range of pressures is:
- 0.0: not touching the screen (hover)
- anything between 0 and 1: very light touch
- 1.0: touching the screen at a "normal" level (tap)
- anything above 1.0: force pushing to various degrees
- maxPressure: the maximum that the device can detect

If the OS has a different concept of "normal" then the data should be scaled to fit that range by the engine.

The meanings of values between 0 and 1 and above 1 are never going to be very consistent, even with the same hardware, user to user, since different hands have different strength muscles.

I think we could have the gesture recognizer normalize the range further, e.g. 0.0-1.0 for values less than 1.0 and 1.0-2.0 for values above 1.0, and base it on the normalized range (for most devices, the value will always be 1.0). It won't be perfect, but it's not clear to me what a developer would do that could be better even if we provide raw values to a predicate.

File: packages/flutter/lib/src/gestures/force_press.dart (hunk 25 above)
Comment (xster) [reply packages/flutter/lib/src/gestures/force_press.dart]: I would write:
```
However, if the ForcePressGestureRecognizer is the only recognizer in the arena, thus accepted as soon as the gesture state is possible, the gesture will not yet have started.
```
to connect the relationship

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 30]
```diff
@@ -0,0 +1,262 @@
...
+  }) : assert(globalPosition != null),
+       assert(pressure != null);
+
+  /// The global position at which the function was called.
+  final Offset globalPosition;
+
+  /// The pressure of the pointer on the screen.
+  final double pressure;
+}
+
```
Comment (xster) [inline packages/flutter/lib/src/gestures/force_press.dart]: It's fine to do it in a different PR but I was saying above, there will be cases where users will use the pressure double value before start is triggered. See on the home screen, when you barely force press an icon, it'll visually respond before you trigger the start.

File: packages/flutter/lib/src/gestures/force_press.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (xster) [reply packages/flutter/lib/src/gestures/force_press.dart]: We definitely need to address this before merging. 

To @Hixie's comment. We don't really need to worry about subjectively interpreting the 0-1 range. The Android and iOS APIs already contractually assign meanings to those bounds. We can just pass them directly. 

There are no higher bound API definitions. I would prefer letting start and peak be predicates in those cases. The user may want to curve differently if the device can go to 3 vs if it can go to 10. We don't have that future information and shouldn't lock our users to our present interpretation.

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 31]
```diff
@@ -0,0 +1,262 @@
...
+      final double pressure = _inverseLerp(event.pressureMin, event.pressureMax, event.pressure);
+      _lastPosition = event.position;
+      _lastPressure = pressure;
+
+      if (_state == _ForceState.possible) {
+        if (pressure > startPressure) {
+          _state = _ForceState.started;
+          resolve(GestureDisposition.accepted);
+        } else if (event.delta.distanceSquared > kTouchSlop)
+          resolve(GestureDisposition.rejected);
```
Comment (xster) [inline packages/flutter/lib/src/gestures/force_press.dart]: if you have brackets for if, have brackets for else too

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 32]
```diff
@@ -0,0 +1,262 @@
...
+    stopTrackingPointer(pointer);
+    didStopTrackingLastPointer(pointer);
+  }
+
+  double _inverseLerp(double min, double max, double t) {
+    return (t - min) / (max - min);
+  }
+
+  @override
+  String get debugDescription => 'forcepress';
```
Comment (xster) [inline packages/flutter/lib/src/gestures/force_press.dart]: 2 words?

**Review by xster — COMMENTED** [review]
The rest looks good but we need to solve the range issue.

File: packages/flutter/lib/src/gestures/events.dart [hunk 33]
```diff
@@ -334,6 +334,7 @@ class PointerAddedEvent extends PointerEvent {
     int device = 0,
     Offset position = Offset.zero,
     bool obscured = false,
+    double pressure = 0.0,
```
[the remark this answers is not in the document]
Comment (jslavitz) [reply packages/flutter/lib/src/gestures/events.dart]: So Ian told me that added/removed/canceled events should have pressure 0 because there's technically no touch occurring.

File: packages/flutter/lib/src/gestures/events.dart [hunk 34]
```diff
@@ -567,7 +573,7 @@ class PointerUpEvent extends PointerEvent {
     Offset position = Offset.zero,
     int buttons = 0,
     bool obscured = false,
-    double pressure = 1.0,
+    double pressure = 0.0,
```
[the remark this answers is not in the document]
Comment (jslavitz) [reply packages/flutter/lib/src/gestures/events.dart]: Doesn't it make sense for the pressure to be 0 for hover?

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 35]
```diff
@@ -0,0 +1,293 @@
...
+/// [ForcePressGestureRecognizer.startPressure]. A long press recognizer will
+/// win when the press down time exceeds the threshold time as long as the
+/// pointer's pressure was never greater than
+/// [ForcePressGestureRecognizer.startPressure] in that duration.
+///
+/// As of November, 2018 iPhone devices of generation 6S and higher have
+/// force touch functionality, with the exception of the iPhone XR. In addition,
+/// a small handful of Android devices have this functionality as well.
+///
+/// Reported pressure will always be in the range [0.0, 1.0], where 1.0 is
```
Comment (xster) [inline packages/flutter/lib/src/gestures/force_press.dart]: Reference the interpolation here too

File: packages/flutter/lib/src/gestures/force_press.dart [hunk 36]
```diff
@@ -0,0 +1,293 @@
...
+/// maximum pressure and 0.0 is minimum pressure.
+class ForcePressGestureRecognizer extends OneSequenceGestureRecognizer {
+  /// Creates a force press gesture recognizer.
+  ///
+  /// The [startPressure] defaults to 0.4, and [peakPressure] defaults to 0.85
+  /// where a value of 0.0 is no pressure and a value of 1.0 is maximum pressure.
+  ///
+  /// [startPressure], [peakPressure] and [interpolation] must not be null.
+  /// [peakPressure] must be greater than [startPressure]. [interpolation] must
+  /// always return a value in the range [0.0, 1.0] for
```
Comment (xster) [inline packages/flutter/lib/src/gestures/force_press.dart]: This sentence is incomplete

File: packages/flutter_test/lib/src/test_pointer.dart [hunk 37]
```diff
@@ -43,12 +43,31 @@ class TestPointer {
...
+    return isDown;
+  }
+
   /// Create a [PointerDownEvent] at the given location.
   ///
   /// By default, the time stamp on the event is [Duration.zero]. You
   /// can give a specific time stamp by passing the `timeStamp`
   /// argument.
-  PointerDownEvent down(Offset newLocation, { Duration timeStamp = Duration.zero }) {
+  PointerDownEvent down(Offset newLocation, { Duration timeStamp = Duration.zero, }) {
```
Comment (xster) [inline packages/flutter_test/lib/src/test_pointer.dart]: revert this line

File: packages/flutter/test/gestures/force_press_test.dart [hunk 38]
```diff
@@ -0,0 +1,345 @@
...
+
+    tester.route(pointer.up());
+
+    expect(started, 1);
+    expect(peaked, 0);
+    expect(updated, 1);
+    expect(ended, 1);
+    expect(didStartPan, 0);
+  });
+}
```
Comment (xster) [inline packages/flutter/test/gestures/force_press_test.dart]: Test the interpolation mechanism

**Review by xster — APPROVED** [review]

**Comment by xster** [comment]
\o/

**Comment by xster** [comment]
@Hixie understood. Will be addressed in https://github.com/flutter/flutter/issues/26563#issuecomment-454935759
